### PR TITLE
Fix ACO and add try except in fit

### PR DESCRIPTION
**Pull request by RomanSraj**

### CODE REVIEW

File: suprb/optimizer/solution/aco/builder.py [hunk 1]
```diff
@@ -90,14 +90,14 @@ def relative_bounds_overlap(A: Rule, B: Rule) -> np.array:
     If one rule lies completely within the bounds of the other rule, the overlap is 1.
     """
 
-    intersections = (A.bounds[:, 0] <= B.bounds[:, 1]) & (B.bounds[:, 0] <= A.bounds[:, 1])
+    intersections = (A.input_space[:, 0] <= B.input_space[:, 1]) & (B.input_space[:, 0] <= A.input_space[:, 1])
```
Comment (heidmic) [inline suprb/optimizer/solution/aco/builder.py:93]: I dont think it is the input space that is referenced here. should be the actual matching function (I think)

File: suprb/suprb.py [hunk 2]
```diff
@@ -147,15 +147,27 @@ def fit(self, X: np.ndarray, y: np.ndarray, cleanup=False):
 
         # Fill population before first step
         if self.n_initial_rules > 0:
-            self._discover_rules(X, y, self.n_initial_rules)
+            try:
+                self._discover_rules(X, y, self.n_initial_rules)
+            except:
+                warnings.warn("Error in rule discovery first iteration")
```
Comment (heidmic) [inline suprb/suprb.py:153]: Change to:
"An error has occured when trying to discover rules for the first time. This is likely due to a bad configuration".

Additionally, please include the original error message

File: suprb/suprb.py [hunk 3]
```diff
@@ -147,15 +147,27 @@ def fit(self, X: np.ndarray, y: np.ndarray, cleanup=False):
...
+                return self
 
         # Main loop
         for self.step_ in range(self.n_iter):
             # Insert new rules into population
-            self._discover_rules(X, y, self.n_rules)
+            try:
+                self._discover_rules(X, y, self.n_rules)
+            except:
+                warnings.warn("Error in rule discovery first iteration")
```
Comment (heidmic) [inline suprb/suprb.py:162]: Change to:
"An error has occured when trying to discover rules." 

Additionally, please include the original error message

File: suprb/suprb.py [hunk 4]
```diff
@@ -147,15 +147,27 @@ def fit(self, X: np.ndarray, y: np.ndarray, cleanup=False):
...
+                return self
 
             # Optimize solutions
-            self._compose_solution(X, y)
+            try:
+                self._compose_solution(X, y)
+            except:
+                warnings.warn("Error in rule discovery first iteration")
```
Comment (heidmic) [inline suprb/suprb.py:169]: Change to:
"An error has occured when trying to compose a solution."

Additionally, please include the original error message

File: suprb/suprb.py (hunk 3 above)
Comment (heidmic) [reply suprb/suprb.py:162]: Can we include the iteration without changing the for loop?

File: suprb/suprb.py [hunk 5]
```diff
@@ -147,15 +147,27 @@ def fit(self, X: np.ndarray, y: np.ndarray, cleanup=False):
 
         # Fill population before first step
         if self.n_initial_rules > 0:
-            self._discover_rules(X, y, self.n_initial_rules)
+            try:
+                self._discover_rules(X, y, self.n_initial_rules)
+            except:
+                warnings.warn("Error in rule discovery first iteration")
+                return self
```
Comment (heidmic) [inline suprb/suprb.py:154]: what is the fitness of the returned model? it should be horrificly bad

**Review by heidmic — CHANGES_REQUESTED** [review]

File: suprb/suprb.py (hunk 5 above)
Comment (RomanSraj) [reply suprb/suprb.py:157]: Changed it, so in predict we get a very bad MSE (almost 1, something like 0.99999) when we enter an error case

File: suprb/suprb.py (hunk 3 above)
Comment (RomanSraj) [reply suprb/suprb.py]: I think we can, but I don't want to mess this up now. Maybe look into this if we want to refactor the code as a whole to have a clean code throughout the project?

**Review by heidmic — APPROVED** [review]

**Pull request merged**